include/fast_float/ascii_number.h

```diff
@@ -26,13 +27,13 @@ fastfloat_really_inline constexpr uint64_t byteswap(uint64_t val) {
     | (val & 0x000000000000FF00) << 40
     | (val & 0x00000000000000FF) << 56;
 }
-
+template <typename UC>
 fastfloat_really_inline FASTFLOAT_CONSTEXPR20
-uint64_t read_u64(const char *chars) {
-  if (cpp20_and_in_constexpr()) {
-    uint64_t val = 0;
+uint64_t read_u64(UC const * chars) {
+  if (cpp20_and_in_constexpr() || sizeof(UC) > 1) {
+    uint64_t val{};
     for(int i = 0; i < 8; ++i) {
-      val |= uint64_t(*chars) << (i*8);
+      val |= uint64_t(char(*chars)) << (i * 8);
```

Comment:
This ignores upper bytes for char16_t/char32_t.
is_made_of_eight_digits_fast() will be inaccurate as it checks only the lowest byte of each char16_t/char32_t.

Reply:
I agree. Good catch. We need to check that the ignored bytes are zeroes.

Reply:
I have fixed this in a commit. We want do this optimization for UTF-16 and UTF-32.

Reply:
@Pharago Why should the higher bytes be ignored? they are relevant in knowing if the char16_t/char32_t is a digit or not. For eg. with char16_t, 0x0030 is a digit (i.e. '0'), but 0x0130 ('İ') is not.

Reply:
@mayawarrier yes, you are absolutely right, letting the whole UC be or'd to val will cause the test to fail if there is anything but 0 on the other bytes

Reply:
I agree with @mayawarrier : we cannot dismiss the bytes, we must check that they are zero... and that requires some work. What I did for now is to just move the issue aside and 'de-optimize' the UTF-16 and UTF-32 code. This can be re-optimized later... but it needs to be done safely.

It is pretty easy with SIMD to check that you are dealing with ASCII-as-UTF-16, and to prune the zero bytes. But I am concerned that it is not likely to be fast with scalar code.

Reply:
@mayawarrier it still leaves the first non zero digit i.e: 0x0130 ('İ') be shifted until overflows and becomes 0x30, so L"İ0000000" is still a valid is_made_of_eight_digits_fast

Reply:
@Pharago This code no longer appears, and I suggest we process it as a new PR. We want to keep the PRs small and easily understood. I think we can put back this work separately, accompanied by benchmarks and additional tests.

Right now, we do not test for arbitrary Unicode strings. So we have to keep the code simple otherwise we risk introducing nasty bugs.

Reply:
@lemire I think I should close #192 and make a proper one once it passes all the tests, what do you think?

Reply:
@Pharago I would like to merge this PR (https://github.com/fastfloat/fast_float/pull/197). Of course, you’d get full credit. I can add a co-written commit so your name will be in the logs. Also I am adding you to the contributors. Then other PRs could add further optimizations.

Reply:
@lemire Great, thank you, I'll keep working on my develop branch.